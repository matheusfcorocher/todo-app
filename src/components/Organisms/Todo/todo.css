body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}

.todo-panel {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  width: calc(100vw - 2rem);
  max-width: 30rem;
  padding: 0.5rem;
  margin: 0 1rem 0 1rem;
  gap: 0.5rem;
  box-shadow: #403539 0px 0px 0px 1px inset;
  background: #6f5a62;
  border-radius: 0.2rem;
}