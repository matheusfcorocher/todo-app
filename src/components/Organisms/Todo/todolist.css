body {
  margin: 0;
  font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', 'Roboto', 'Oxygen',
    'Ubuntu', 'Cantarell', 'Fira Sans', 'Droid Sans', 'Helvetica Neue',
    sans-serif;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
}

code {
  font-family: source-code-pro, Menlo, Monaco, Consolas, 'Courier New',
    monospace;
}

.todo-panel {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  max-width: 30rem;
  padding: 0.5rem;
  gap: 0.5rem;
  box-shadow: #00000026 0px 0px 0px 1px inset;
}