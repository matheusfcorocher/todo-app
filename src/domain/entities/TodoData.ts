import { v4 as uuidv4 } from 'uuid';
import { isStringBlank } from '../../lib/string';

export type TodoData = {
    id: string;
    title: string;
    isCompleted: boolean;
}

export type Todos = Array<TodoData>;

type AddTodoParameters = {
    todos: Todos,
    title: string
};

export function addTodo({todos, title} : AddTodoParameters): Todos {
    const newTodo: TodoData = {
      id: uuidv4(),
      title,
      isCompleted: false
    };
    return [...todos, newTodo];
}

type DeleteTodoParameters = {
    todos: Todos,
    id: string
};

export function deleteTodo({todos, id} :DeleteTodoParameters): Todos {
    const newTodos = todos.filter((todo) => todo.id !== id);
    return newTodos;
}

type UpdateTodoParameters = {
    todos: Todos,
    id: string,
    newTitle: string
};

export function updateTodoTitle({todos, id, newTitle}: UpdateTodoParameters): Todos {
    if (isStringBlank(newTitle)) {
      const newTodos = deleteTodo({todos, id});
      return newTodos;
    }

    const oldTodo = todos.find((todo) => todo.id === id);
    if (oldTodo) {
        const index = todos.findIndex((todo) => todo.id === id);
        const newTodo: TodoData = { ...oldTodo, title: newTitle };
        const newTodos = [...todos.slice(0, index), newTodo, ...todos.slice(index + 1)];
        return(newTodos);
    }

    return todos;
}

type CompleteAllTodosParameters = {
    todos: Todos
};

export function completeAllTodosItem({todos} : CompleteAllTodosParameters): Todos {
    const newTodos = todos.map(todo => {
      return {
        ...todo,
        isCompleted: true
      }
    });
    return newTodos;
  }

