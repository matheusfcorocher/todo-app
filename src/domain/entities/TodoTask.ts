import { v4 as uuidv4 } from 'uuid';
import { isStringBlank } from '../../lib/string';

//Entity

export type TodoTask = {
    id: string;
    title: string;
    isCompleted: boolean;
}

export type TodoTasks = Array<TodoTask>;

//Functions

export type AddTodoTaskParameters = {
    todoTasks: TodoTasks,
    title: string
};

export function addTodoTask({todoTasks, title} : AddTodoTaskParameters): TodoTasks {
    const newTodo: TodoTask = {
      id: uuidv4(),
      title,
      isCompleted: false
    };
    return [...todoTasks, newTodo];
}

export type DeleteTodoTaskParameters = {
    todoTasks: TodoTasks,
    id: string
};

export function deleteTodoTask({todoTasks, id} :DeleteTodoTaskParameters): TodoTasks {
    const newTodoTasks = todoTasks.filter((todo) => todo.id !== id);
    return newTodoTasks;
}

export type UpdateTodoTaskTitleParameters = {
    todoTasks: TodoTasks,
    id: string,
    newTitle: string
};

export function updateTodoTaskTitle({todoTasks, id, newTitle}: UpdateTodoTaskTitleParameters): TodoTasks {
    if (isStringBlank(newTitle)) {
      const newTodoTasks = deleteTodoTask({todoTasks, id});
      return newTodoTasks;
    }

    const oldTodo = todoTasks.find((todo) => todo.id === id);
    if (oldTodo) {
        const index = todoTasks.findIndex((todo) => todo.id === id);
        const newTodo: TodoTask = { ...oldTodo, title: newTitle };
        const newTodoTasks = [...todoTasks.slice(0, index), newTodo, ...todoTasks.slice(index + 1)];
        return(newTodoTasks);
    }

    return todoTasks;
}

export type UpdateTodoTaskStateParameters = {
    todoTasks: TodoTasks,
    id: string,
    state: boolean
};

export function updateTodoTaskState({todoTasks, id, state} : UpdateTodoTaskStateParameters): TodoTasks {
    const oldTodo = todoTasks.find((todo) => todo.id === id);
    if (oldTodo) {
      const index = todoTasks.findIndex((todo) => todo.id === id);
      const newTodo: TodoTask = { ...oldTodo, isCompleted: state };
      const newTodoTasks = [...todoTasks.slice(0, index), newTodo, ...todoTasks.slice(index + 1)];
      return newTodoTasks;
    }
    return todoTasks;
}

export type CompleteAllTodoTasksParameters = {
    todoTasks: TodoTasks
};

export function completeAllTodoTasks({todoTasks} : CompleteAllTodoTasksParameters): TodoTasks {
    const newTodoTasks = todoTasks.map(todo => {
      return {
        ...todo,
        isCompleted: true
      }
    });
    return newTodoTasks;
}

type IncompleteAllTodoTasksParameters = {
    todoTasks: TodoTasks
};

export function incompleteAllTodoTasks({todoTasks} : IncompleteAllTodoTasksParameters): TodoTasks {
    const newTodoTasks = todoTasks.map(todo => {
      return {
        ...todo,
        isCompleted: false
      }
    });
    return newTodoTasks;
  }

