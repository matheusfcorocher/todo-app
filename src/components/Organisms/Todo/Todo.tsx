import React, { useState } from 'react';
import { TodoItem } from '../../Molecules/TodoItem/TodoItem';
import '../../../App.css';
import './todo.css';
import TodoMenu from '../../Molecules/TodoMenu/TodoMenu';
import TodoList from '../../Molecules/TodoList/TodoList';
import TodoFooter from '../../Molecules/TodoFooter/TodoFooter';
import { handleCreateTodo } from '../../../App';

export type TodoData = {
  id: string;
  title: string;
  isCompleted: boolean;
}

interface TodoProps {
  todosData?: Array<TodoData>;
  handleCreateTodo: handleCreateTodo;
}

function Todo({ todosData = [], handleCreateTodo }: TodoProps) {
  return (
    <div data-testid="todo" role="group" className={"todo-panel"}>
      <TodoMenu handleCreateTodo={handleCreateTodo} 
      isAllTodosCompleted={false}
      />
      <TodoList todosData={todosData} />
      {todosData.length !== 0 && <TodoFooter todosQuantity={todosData.length} />}
    </div>
  );
}

export default Todo;
