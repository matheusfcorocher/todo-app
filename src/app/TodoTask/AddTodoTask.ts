import { addTodoTask, AddTodoTaskParameters, TodoTasks } from "../../domain/entities/TodoTask";

type MakeAddTodoTaskParameters = {
    storage: ;
}

export function makeAddTodoTask({storage} : MakeAddTodoTaskParameters ): typeof addTodoTask {

    function AddTodoTask({ todoTasks, title }: AddTodoTaskParameters): TodoTasks {
        const newTodosTasks = addTodoTask({ todoTasks, title });
        storage.store({todoTasks: newTodosTasks});
        return newTodosTasks;
    }

    return AddTodoTask;
}