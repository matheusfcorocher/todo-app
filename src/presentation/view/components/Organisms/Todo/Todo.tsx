import React, { useState } from 'react';
import './todo.css';
import TodoMenu from '../../Molecules/TodoMenu/TodoMenu';
import TodoList from '../../Molecules/TodoList/TodoList';
import TodoFooter from '../../Molecules/TodoFooter/TodoFooter';
import { TodoTasks } from '../../../../../domain/entities/TodoTask';
import { TodoTaskControllerReturnType } from '../../../../controllers/TodoTaskController';
import { HandleFilter } from '../../../../../App';
import { todoTaskViewModel } from '../../../../models/TodoTaskViewModel';

interface TodoProps {
  todoTasks?: TodoTasks;
  filter?: boolean;
  todoTaskController: TodoTaskControllerReturnType;
  handleFilter: HandleFilter;
}

function Todo({
  todoTasks = [],
  filter = undefined,
  todoTaskController,
  handleFilter
}: TodoProps) {
  const { isTodoTasksNotEmpty, returnOnlyActiveTodoTasks, isThereAnyTodoTaskCompleted } = todoTaskViewModel;

  return (
    <div data-testid="todo" role="group" className={"todo-panel"}>
      <TodoMenu
        isThereAnyTodoTaskCompleted={isThereAnyTodoTaskCompleted(todoTasks)}
        todoTaskController={todoTaskController}
      />
      {isTodoTasksNotEmpty(todoTasks) &&
        <>
          <TodoList
            todoTasks={todoTasks}
            filter={filter}
            todoTaskController={todoTaskController}
          />
          <TodoFooter
            activeTodoTasksQuantity={returnOnlyActiveTodoTasks(todoTasks).length}
            handleFilter={handleFilter}
            filter={filter}
            todoTaskController={todoTaskController}
            isThereAnyTodoTaskCompleted={isThereAnyTodoTaskCompleted(todoTasks)} />

        </>
      }
    </div>
  );
}

export default Todo;
