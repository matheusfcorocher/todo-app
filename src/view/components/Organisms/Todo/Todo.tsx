import React, { useState } from 'react';
import './todo.css';
import TodoMenu from '../../Molecules/TodoMenu/TodoMenu';
import TodoList from '../../Molecules/TodoList/TodoList';
import TodoFooter from '../../Molecules/TodoFooter/TodoFooter';
import { areThereTodoTasksCompleted, filterTodoTasksByIsCompleted, TodoTasks } from '../../../../domain/entities/TodoTask';
import { HandleFilter } from '../../../../App';
import { TodoTaskControllerReturnType } from '../../../controllers/TodoTaskController';
import { isArrayEmpty } from '../../../../lib/isArrayEmpty/isArrayEmpty';

interface TodoProps {
  todoTasks?: TodoTasks;
  filter?: boolean;
  todoTaskController: TodoTaskControllerReturnType;
  handleFilter: HandleFilter;
}

function Todo({ 
  todoTasks = [],
  filter= undefined, 
  todoTaskController,
  handleFilter
}: TodoProps) {
  const isCompleted = isAllTodosCompleted(todoTasks);

  function isAllTodosCompleted(todoTasks: TodoTasks): boolean {
    const isAllTodosCompleted = todoTasks.find(todo => todo.isCompleted == true);
    if (isAllTodosCompleted) {
      return true;
    }

    return false;
  }

  return (
    <div data-testid="todo" role="group" className={"todo-panel"}>
      <TodoMenu
        isAllTodosCompleted={isCompleted}
        todoTaskController={todoTaskController}
      />
      <TodoList
        todoTasks={todoTasks}
        filter={filter}
        todoTaskController={todoTaskController}
      />
      {!isArrayEmpty(todoTasks) && <TodoFooter 
        todosQuantity={filterTodoTasksByIsCompleted({ todoTasks, isCompleted: false }).length}
        handleFilter={handleFilter}
        todoTaskController={todoTaskController}
        areThereTodoTasksCompleted={areThereTodoTasksCompleted({todoTasks})}      />}
    </div>
  );
}

export default Todo;
