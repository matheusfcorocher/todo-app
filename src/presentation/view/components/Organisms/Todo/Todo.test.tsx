import React from 'react';
import { render, screen } from '@testing-library/react';
import Todo from './Todo';
import { TodoTasks } from '../../../../../domain/entities/TodoTask';

describe("::Components ::Organisms ::Todo", () => {
  test('renders Todo', () => {
    render(<Todo todoTaskController={{
      getTodoTasks(): TodoTasks {
        return [];
      },
      getIsAllTodoTaskCompleted(): boolean {
        return true;
      },
      getIsThereAnyTodoTaskCompleted(): boolean {
        return true;
      },
      getIsTodoTasksNotEmpty(): boolean {
        return true;
      },
      getOnlyActiveTodoTasks(): TodoTasks {
        return [];
      },
      handleAddTodoTask: function (title: string): void {
        throw new Error('Function not implemented.');
      },
      handleDeleteTodoTask: function (id: string): void {
        throw new Error('Function not implemented.');
      },
      handleUpdateTodoTaskTitle: function (id: string, newTitle: string): void {
        throw new Error('Function not implemented.');
      },
      handleUpdateTodoTaskState: function (id: string, state: boolean): void {
        throw new Error('Function not implemented.');
      },
      handleCompleteAllTodoTasks: function (): void {
        throw new Error('Function not implemented.');
      },
      handleIncompleteAllTodoTasks: function (): void {
        throw new Error('Function not implemented.');
      },
      handleDeleteAllCompletedTodoTasks: function (): void {
        throw new Error('Function not implemented.');
      }
    }} isCompletedFilterController={{
      getIsCompletedFilter: function (): boolean | undefined {
        return true;
      },
      getFilterByUrlHash: function (hash: string): boolean | undefined {
        return true;
      },
      handleChangeFilter: function (isCompleted?: boolean | undefined): void {
        throw new Error('Function not implemented.');
      }
    }}/>);
    const todo = screen.getByTestId("todo");
    expect(todo).toMatchSnapshot();
  })
})