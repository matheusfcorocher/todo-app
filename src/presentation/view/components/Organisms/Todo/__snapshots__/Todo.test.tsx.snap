// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`::Components ::Organisms ::Todo renders Todo 1`] = `
<div
  class="todo-panel"
  data-testid="todo"
  role="group"
>
  <div
    class="todo-menu"
    role="group"
  >
    <div
      style="opacity: 0; width: 0px;"
    >
      <button
        class="icon-button complete-button"
        role="button"
        title="Uncheck all todos"
      >
        <svg
          class="list-icon"
          preserveAspectRatio="none"
          role="image"
          viewBox="0 0 24 24"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
            fill=""
          />
        </svg>
      </button>
    </div>
    <input
      class="title-input hidden"
      name="title"
      placeholder="What needs to be done?"
      role="textbox"
      type="text"
      value=""
    />
  </div>
  <ul
    class="todo-list"
    role="list"
  />
  <div
    class="todo-footer"
    role="group"
  >
    <span
      class="todo-count"
    >
      <strong>
        0
        Â 
      </strong>
      item left
    </span>
    <ul
      class="todo-filters"
    >
      <li>
        <a
          class="todo-filter "
          href="#/"
          title="Filter all todo tasks"
        >
          All
        </a>
      </li>
      <li>
        <a
          class="todo-filter "
          href="#/active"
          title="Filter active todo tasks"
        >
          Active
        </a>
      </li>
      <li>
        <a
          class="todo-filter active"
          href="#/completed"
          title="Filter completed todo tasks"
        >
          Completed
        </a>
      </li>
    </ul>
    <button
      class="icon-button small primary gray"
      role="button"
      title="Clear Completed"
    >
      Clear Completed
    </button>
  </div>
</div>
`;
